File: Python/binary_tree_preorder_traversal.py
from utils.examination import Examiner, Testcase
from utils.tree import TreeNode


class Solution:
    def preorderTraversal(self, root: TreeNode|None) -> list[int]:
        if root is None:
            return []
        traversal: list[int] = [root.val]
        traversal.extend(self.preorderTraversal(root.left))
        traversal.extend(self.preorderTraversal(root.right))
        return traversal

def examine()->None:
    Examiner(Solution(), 'preorderTraversal').run((
        Testcase((TreeNode.from_list([1,None,2,3]),), [1,2,3]),
        Testcase((TreeNode.from_list([]),), []),
        Testcase((TreeNode.from_list([1]),), [1]),
        Testcase((TreeNode.from_list([1,2,3,4,5,None,8,None,None,6,7,9]),), [1,2,4,5,6,7,3,8,9]),
    ))

File: TypeScript/solutions/binary-tree-preorder-traversal.ts
// https://leetcode.com/problems/binary-tree-preorder-traversal/description/
import { Examiner, Testcase, TreeNode } from '../utils';


function preorderTraversal(root: TreeNode | null): number[] {
    if (root === null) return [];
    return [root.val, ...preorderTraversal(root.left), ...preorderTraversal(root.right)];
}

export function examine(): void {
    new Examiner(preorderTraversal).run([
        new Testcase([TreeNode.fromList([1, null, 2, 3])], [1, 2, 3]),
        new Testcase([TreeNode.fromList([])], []),
        new Testcase([TreeNode.fromList([1])], [1]),
        new Testcase([TreeNode.fromList([1, 2, 3, 4, 5, null, 8, null, null, 6, 7, 9])], [1, 2, 4, 5, 6, 7, 3, 8, 9]),
    ]);
}
